### PR TITLE
Adding support to autoconfigure Cloud SQL ipTypes

### LINKED ISSUE #2513: Add support to autoconfigure Cloud SQL ipTypes
**Problem: Missing configuration option**
Currently `spring-cloud-gcp-autoconfigure` doesn't support configuring of [ipTypes](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#specifying-ip-types) for [Cloud SQL Socket Factory for JDBC drivers](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory). When the access to a Cloud SQL database is restricted to the `PRIVATE` address the connections will fail since Cloud SQL Socket Factory defaults to the `PUBLIC` address.

**Solution:**
Add missing configuration option.

**Alternatives/workarounds**
* Configure HikariCP
`spring.datasource.hikari.data-source-properties.ipTypes=PRIVATE`
Pro: Works as long as you use the ConnectionPool.
Con: Doesn't work when something else like `org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration` creates a `DataSource` based on the `DataSourceProperties`. In my case that happens since the Flyway migrations are run with another database user than the applications database user, see [FlywayAutoConfiguration#configureDataSource(...)](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration.java#L143).
* For PostgreSQL you can add a `org\postgresql\driverconfig.properties` to your classpath with `ipTypes=PRIVATE`.
Pro: Works for PostgreSQL.
Con: Hiding important configuration a really long way from where people would look for it.

**Proposed code change**
https://github.com/ohardeng/spring-cloud-gcp/tree/autoconfigure-cloud-sql-ip-types is ready to create a pull request from.

### PR DESCRIPTION
**Pull request by ohardeng**
Closes #2513

### CODE REVIEW

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java [hunk 1]
```diff
@@ -161,7 +162,11 @@ public DataSourceProperties cloudSqlDataSourceProperties(
 				LOGGER.warn("Ignoring provided spring.datasource.url. Overwriting it based on the " +
 						"spring.cloud.gcp.sql.instance-connection-name.");
 			}
-			properties.setUrl(cloudSqlJdbcInfoProvider.getJdbcUrl());
+			String jdbcUrl = cloudSqlJdbcInfoProvider.getJdbcUrl();
+			if (StringUtils.hasText(gcpCloudSqlProperties.getIpTypes())) {
+				jdbcUrl = String.format(jdbcUrl + "&ipTypes=%s", gcpCloudSqlProperties.getIpTypes());
```
Comment (meltsufin) [inline spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java:167]: Have you considered putting this logic into `DefaultCloudSqlJdbcInfoProvider` instead?
That class is specifically designed for creating the JDBC URL string.

File: docs/src/main/asciidoc/sql.adoc [hunk 2]
```diff
@@ -64,6 +64,7 @@ There is also some configuration specific to Google Cloud SQL:
 | `spring.cloud.gcp.sql.database-name` | Name of the database to connect to. |
 | `spring.cloud.gcp.sql.instance-connection-name` | A string containing a Google Cloud SQL instance's project ID, region and name, each separated by a colon.
 For example, `my-project-id:my-region:my-instance-name`. |
+| `spring.cloud.gcp.sql.ip-types` | Allows you to specify a comma delimited list of preferred IP types for connecting to a Cloud SQL instance. See https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#specifying-ip-types[Cloud SQL Socket Factory - Specifying IP Types] |
```
Comment (meltsufin) [inline docs/src/main/asciidoc/sql.adoc:67]: Can you mention the default `PUBLIC,PRIVATE` here?

**Review by meltsufin — CHANGES_REQUESTED** [review]
@ohardeng Thank you for such a nice PR!
I just have a couple of minor polish comments.

File: spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java (hunk 1 above)
Comment (ohardeng) [reply spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java]: No. I remember that I looked at patching the `jdbcUrlTemplate` in `DatabaseType`, but I didn't have the `GcpCloudSqlProperties` there. I must have overlooked `DefaultCloudSqlJdbcInfoProvider`. 
Code moved.

File: docs/src/main/asciidoc/sql.adoc (hunk 2 above)
Comment (ohardeng) [reply docs/src/main/asciidoc/sql.adoc]: Done

Comment (meltsufin) [reply docs/src/main/asciidoc/sql.adoc]: If you look at the preview, there is also a "Default value" column you can populate.
image

Comment (ohardeng) [reply docs/src/main/asciidoc/sql.adoc]: Yes, I saw the "Default value" column, but felt that populating that would signify that spring-cloud-gcp was setting thoose default values. I now see that `spring.cloud.gcp.sql.credentials.location` specifies that it's Spring GCP Boot starter that supplying the defaults.

Suggested rewrite:
| `spring.cloud.gcp.sql.ip-types` | Allows you to specify a comma delimited list of preferred IP types for connecting to a Cloud SQL instance. See https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#specifying-ip-types[Cloud SQL Socket Factory - Specifying IP Types] |  Default values (`PUBLIC,PRIVATE`) provided by Cloud SQL Socket Factory

Comment (meltsufin) [reply docs/src/main/asciidoc/sql.adoc]: The default doesn't have to be provided by this project. The column is there to make it more clear what would be the default.

File: docs/src/main/asciidoc/sql.adoc [hunk 3]
```diff
@@ -64,6 +64,7 @@ There is also some configuration specific to Google Cloud SQL:
 | `spring.cloud.gcp.sql.database-name` | Name of the database to connect to. |
 | `spring.cloud.gcp.sql.instance-connection-name` | A string containing a Google Cloud SQL instance's project ID, region and name, each separated by a colon.
 For example, `my-project-id:my-region:my-instance-name`. |
+| `spring.cloud.gcp.sql.ip-types` | Allows you to specify a comma delimited list of preferred IP types for connecting to a Cloud SQL instance. Left unconfigured Cloud SQL Socket Factory will default it to `PUBLIC,PRIVATE`. See https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#specifying-ip-types[Cloud SQL Socket Factory - Specifying IP Types] |
```
Comment (meltsufin) [inline docs/src/main/asciidoc/sql.adoc:67]:
```suggestion
| `spring.cloud.gcp.sql.ip-types` | Allows you to specify a comma delimited list of preferred IP types for connecting to a Cloud SQL instance. Left unconfigured Cloud SQL Socket Factory will default it to `PUBLIC,PRIVATE`. See https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory#specifying-ip-types[Cloud SQL Socket Factory - Specifying IP Types] | `PUBLIC,PRIVATE`
```

**Review by meltsufin — APPROVED** [review]
I just added the default to the column. Ready to merge.
Thanks again for the PR!

**Pull request merged**